### PR TITLE
ARROW-4960: [R] Build r-arrow conda package in crossbow

### PR DESCRIPTION
**Pull request by xhochy**
FYI @nealrichardson 

This probably won't work on first try but I hope to get fixes in before the release. I fear breakage with `cpp11` and `clang`-on-windows.

### CODE REVIEW

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu

**Comment by nealrichardson** [comment]
> I fear breakage with `cpp11` and `clang`-on-windows.

`cpp11` is header only, hopefully that helps?

**Comment by xhochy** [comment]
> > I fear breakage with `cpp11` and `clang`-on-windows.
> 
> `cpp11` is header only, hopefully that helps?

Actually makes it worse as we need to compile everything of it with clang and clang is quite strict on Windows about everything.

**Comment by xhochy** [comment]
I also need to add it to the conda recipe but CI will probably fail much earlier.

File: dev/tasks/conda-recipes/r-arrow/configure.win [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+echo "PKG_CPPFLAGS=-DNDEBUG -I\"${LIBRARY_PREFIX}/include\" -I\"${PREFIX}/include\" -DARROW_R_WITH_ARROW" > src/Makevars.win
```
Comment (nealrichardson) [inline dev/tasks/conda-recipes/r-arrow/configure.win:5]: assuming arrow-cpp is built with `ARROW_S3=ON`, do this:

```suggestion
echo "PKG_CPPFLAGS=-DNDEBUG -I\"${LIBRARY_PREFIX}/include\" -I\"${PREFIX}/include\" -DARROW_R_WITH_ARROW -DARROW_R_WITH_S3" > src/Makevars.win
```

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu

**Comment by xhochy** [comment]
@kszucs Do you have an idea why my new `r_config` parameter doesn't work? I get the following traceback by the bot:

```
Cloning into '/tmp/tmpzsnkz8m4/arrow'...
From https://github.com/apache/arrow
 * [new ref]             refs/pull/8371/head -> ARROW-4960
Switched to branch 'ARROW-4960'
Cloning into '/tmp/tmpzsnkz8m4/crossbow'...
Traceback (most recent call last):
  File "/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py", line 720, in render_files
    rendered = template.render(task=self, **params)
  File "/usr/share/miniconda/lib/python3.7/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/share/miniconda/lib/python3.7/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/share/miniconda/lib/python3.7/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 9, in top-level template code
jinja2.exceptions.UndefinedError: 'r_config' is undefined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py", line 1653, in <module>
    crossbow(obj={}, auto_envvar_prefix='CROSSBOW')
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/share/miniconda/lib/python3.7/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py", line 1439, in submit
    config.validate()
  File "/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py", line 1063, in validate
    queue_remote_url='https://github.com/org/crossbow'
  File "/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py", line 724, in render_files
    path, e.__class__.__name__, str(e)
RuntimeError: Failed to render template `/tmp/tmpzsnkz8m4/arrow/dev/tasks/conda-recipes/azure.linux.yml` with UndefinedError: 'r_config' is undefined
Command '['/tmp/tmpzsnkz8m4/arrow/dev/tasks/crossbow.py', '--queue-path', '/tmp/tmpzsnkz8m4/crossbow', '--output-file', '/home/runner/work/arrow/arrow/result.yml', 'submit', '--job-prefix', 'actions', '--arrow-remote', 'https://github.com/xhochy/arrow.git', '--arrow-branch', 'ARROW-4960', 'conda-linux-gcc-py36-cpu']' returned non-zero exit status 1.
```

**Comment by kszucs** [comment]
@xhochy you need to add the `r_config` param to all tasks using the same yml template. If you wan't to set it for just specific builds then handle it as an optional variable from jinja.

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-linux-gcc-py36-cpu

File: dev/tasks/conda-recipes/r-arrow/build_win.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+set -exuo pipefail
+
+# Patch Rcpp
+# https://github.com/RcppCore/Rcpp/pull/1069
+sed -i -e 's; abs; std::abs;g' $BUILD_PREFIX/Lib/R/library/Rcpp/include/Rcpp/DataFrame.h
```
Comment (nealrichardson) [inline dev/tasks/conda-recipes/r-arrow/build_win.sh:7]:
```suggestion
```

File: dev/tasks/conda-recipes/r-arrow/build_win.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+# Rename arrow.dll to lib_arrow.dll to avoid conflicts with the arrow-cpp arrow.dll
+sed -i -e 's/R_init_arrow/R_init_lib_arrow/g' r/src/arrowExports.cpp
+sed -i -e 's/useDynLib(arrow/useDynLib(lib_arrow/g' r/NAMESPACE
+
+# Make exported symbols visible
+# This needs to be patched in https://github.com/wch/r-source/blob/trunk/src/include/R_ext/Visibility.h
+sed -i -e 's/attribute_visible/__declspec(dllexport)/g' $BUILD_PREFIX/Lib/R/library/Rcpp/include/RcppCommon.h
```
Comment (nealrichardson) [inline dev/tasks/conda-recipes/r-arrow/build_win.sh:15]:
```suggestion
```

**Review by nealrichardson — COMMENTED** [review]
Here's some more Rcpp cleanup. Do we test the windows build on arrow's CI?

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-win-vs2017-py36 conda-osx-clang-py37

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-win-vs2017-py36

**Comment by xhochy** [comment]
Windows fails with 
```
In file included from arrowExports.cpp:2:
In file included from D:/bld/r-arrow_1602163814614/_h_env/Lib/R/library/cpp11/include\cpp11.hpp:5:
In file included from D:/bld/r-arrow_1602163814614/_h_env/Lib/R/library/cpp11/include\cpp11/as.hpp:5:
In file included from C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\memory:6:
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\xmemory:217:75: error: indirection requires pointer operand ('const cpp11::r_vector<int>::const_iterator' invalid)
                        _Uses_default_construct_t<_Alloc, decltype(_Unfancy(_UDest)), decltype(*_UFirst)>()));
                                                                                               ^~~~~~~~
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\vector:1823:11: note: in instantiation of function template specialization 'std::_Uninitialized_copy<cpp11::r_vector<int>::const_iterator, int *, std::allocator<int> >' requested here
                return (_Uninitialized_copy(_First, _Last, _Dest, this->_Getal()));
                        ^
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\vector:708:22: note: in instantiation of function template specialization 'std::vector<int, std::allocator<int> >::_Ucopy<cpp11::r_vector<int>::const_iterator>' requested here
                        this->_Mylast() = _Ucopy(_First, _Last, this->_Myfirst());
                                          ^
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\vector:723:3: note: in instantiation of function template specialization 'std::vector<int, std::allocator<int> >::_Range_construct_or_tidy<cpp11::r_vector<int>::const_iterator>' requested here
                _Range_construct_or_tidy(_Get_unwrapped(_First), _Get_unwrapped(_Last), _Iter_cat_t<_Iter>{});
                ^
D:/bld/r-arrow_1602163814614/_h_env/Lib/R/library/cpp11/include\cpp11/r_vector.hpp:945:10: note: in instantiation of function template specialization 'std::vector<int, std::allocator<int> >::vector<cpp11::r_vector<int>::const_iterator, void>' requested here
  return {obj.begin(), obj.end()};
         ^
././arrow_cpp11.h:217:57: note: in instantiation of function template specialization 'cpp11::as_cpp<std::vector<int, std::allocator<int> >, int>' requested here
  explicit ConstReferenceInput(SEXP from) : obj_(cpp11::as_cpp<T>(from)) {}
                                                        ^
arrowExports.cpp:4411:49: note: in instantiation of member function 'arrow::r::ConstReferenceInput<std::vector<int, std::allocator<int> > >::ConstReferenceInput' requested here
        arrow::r::Input<const std::vector<int>&>::type column_indices(column_indices_sexp);
                                                       ^
1 error generated.

```

**Comment by nealrichardson** [comment]
@bkietz can you take a look at that? Windows conda uses MSVC so it wouldn't surprise me if there were some corners in `cpp11` that weren't robust to that.

**Comment by xhochy** [comment]
It's actually clang-with-MSVC headers ;)

**Comment by bkietz** [comment]
unfortunately `cpp11::r_vector::const_iterator::operator*` is not marked const, but MSVC's impl of `std::copy` marks its [arguments const](https://github.com/microsoft/STL/blob/3d88a64538827700e50c757877eb8413338ec281/stl/inc/xmemory#L1557). We'd probably be able to workaround this in `*Input`. I'll open an issue upstream

**Comment by xhochy** [comment]
I can happily report that the iterator is the only issue and the build passes locally for me when I add a `const` at the end of the declaration and to all implementations.

**Comment by nealrichardson** [comment]
Great! Can we apply that patch to https://github.com/conda-forge/r-cpp11-feedstock, and then this will work?

**Comment by xhochy** [comment]
> Great! Can we apply that patch to https://github.com/conda-forge/r-cpp11-feedstock, and then this will work?

Yes!

**Comment by xhochy** [comment]
See https://github.com/conda-forge/r-cpp11-feedstock/pull/4

**Comment by xhochy** [comment]
We can trigger crossbow in ~2h here.

**Comment by xhochy** [comment]
@github-actions crossbow submit -g conda

**Comment by xhochy** [comment]
@nealrichardson Can you have a look at the R error on Windows? Do you have an idea?

**Comment by nealrichardson** [comment]
> @nealrichardson Can you have a look at the R error on Windows? Do you have an idea?

https://dev.azure.com/ursa-labs/crossbow/_build/results?buildId=18698&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=41795ef0-6501-5db4-3ad4-33c0cf085626&l=1403 

https://github.com/apache/arrow/blob/master/r/R/arrowExports.R#L1592

I think this is just the first C++ function being called, not something special about it. So that suggests to me that something's wrong about r/src/arrowExports.cpp, and since this is windows only, perhaps something is off with the `sed`s in build_win.sh

**Comment by nealrichardson** [comment]
@xhochy one other suggestion: if this patch means we'll now have nightly conda R packages, can you edit the language in r/README.md and r/vignettes/install.Rmd about nightlies (and the lack of nightlies for conda)? And ideally edit https://github.com/apache/arrow/blob/master/r/R/install-arrow.R#L58-L60 in install_arrow() to point to the nightly conda repo too.

**Comment by xhochy** [comment]
Ah, I guess I need to some different magic again to export the C++ symbols. On Windows/MSVC they need to be explicitly exported in contrary to all other toolchains.

**Comment by xhochy** [comment]
@nealrichardson The issue is that `cpp11` doesn't have an equivalent to `RcppExport` which was used in the past. This is required for MSVC and clang-on-MSVC and a nice-to-have on Unixes.

CMake has an alternative way to produce these exports using `WINDOWS_EXPORT_ALL_SYMBOLS=ON` but we aren't using CMake here and I'm unable to find in the CMake sources what they are actually doing.

**Comment by bkietz** [comment]
@xhochy @nealrichardson I'm probably misunderstanding something but I'm somewhat surprised we need the symbols to be visible; `arrowExports.cpp` includes an explicit lookup table:

```c++
static const R_CallMethodDef CallEntries[] = {
// ...
		{ "_arrow_all_record_batches", (DL_FUNC) &_arrow_all_record_batches, 1}, 
		{ "_arrow_Table__from_record_batches", (DL_FUNC) &_arrow_Table__from_record_batches, 2}, 
		{ "_arrow_Table__from_dots", (DL_FUNC) &_arrow_Table__from_dots, 2}, 
		{ "_arrow_GetCpuThreadPoolCapacity", (DL_FUNC) &_arrow_GetCpuThreadPoolCapacity, 0}, 
		{ "_arrow_SetCpuThreadPoolCapacity", (DL_FUNC) &_arrow_SetCpuThreadPoolCapacity, 1}, 
		{NULL, NULL, 0}
};

extern "C" void R_init_arrow(DllInfo* dll){
  R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
  R_useDynamicSymbols(dll, FALSE);
}
```

**Comment by xhochy** [comment]
I'm also a bit confused why this is necessary but it was already the same situation with Rcpp and there things only worked once I've added `__declspec(dllexport)` to all the generated functions.

File: dev/tasks/conda-recipes/r-arrow/build_win.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+set -exuo pipefail
+
+
+# Rename arrow.dll to lib_arrow.dll to avoid conflicts with the arrow-cpp arrow.dll
+sed -i -e 's/R_init_arrow/R_init_lib_arrow/g' r/src/arrowExports.cpp
```
Comment (bkietz) [inline dev/tasks/conda-recipes/r-arrow/build_win.sh:7]:
```suggestion
sed -i -e 's/void R_init_arrow/__declspec(dllexport) void R_init_lib_arrow/g' r/src/arrowExports.cpp
```
Maybe lookup of the *init* is failing silently, then falling back to dynamic lookup

**Comment by xhochy** [comment]
@github-actions crossbow submit conda-win-vs2017-py36

File: dev/tasks/conda-recipes/r-arrow/build_win.sh (hunk 4 above)
Comment (xhochy) [reply dev/tasks/conda-recipes/r-arrow/build_win.sh]: Massive! This worked.

**Comment by xhochy** [comment]
@github-actions crossbow submit -g conda

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Great, looks like this is passing now. Before we merge, can you please:

* Update/add documentation about the conda builds to help with future maintenance. Even my relatively naive eyes notice some non-obvious things, like how it appears that we build for R 3.6 on the "py36" jobs and R 4.0 on the "py37" jobs--and I'm sure there are more subtleties that I'm not aware of. 
* Remove the test-conda-r-4.0 job--do we need it anymore? It seems that it's testing that you can build the R package from source using conda as a dependency source, but that seems more than covered by these new packaging jobs
* Update the R documentation and install_arrow to reflect that there are conda nightlies (there will be conda nightlies for R after this, right?)

**Comment by nealrichardson** [comment]
@xhochy what's the status of this? Can we get this merged soon?

**Comment by xhochy** [comment]
@nealrichardson Can you point me to the places where you want to have the two "documentation" bullet points from above written down? I'll add them then.

**Comment by xhochy** [comment]
@github-actions crossbow submit -g conda

**Comment by nealrichardson** [comment]
> Update/add documentation about the conda builds to help with future maintenance. 

I have no idea if/where we have this kind of documentation about the nightly builds, but we should have some. Maybe it belongs in `dev/README.md` or linked to there? If there is none, feel free to open a JIRA about it (and include a note in the JIRA about the connection between the R and pyarrow conda builds so we can document it there).

> Update the R documentation and install_arrow to reflect that there are conda nightlies

* https://github.com/apache/arrow/blob/master/r/README.md#installing-a-development-version
* https://github.com/apache/arrow/blob/master/r/vignettes/install.Rmd#L44
* https://github.com/apache/arrow/blob/master/r/R/install-arrow.R#L58-L59 make the outer check `if (conda)` and then inside that `if (nightly) { ... } else { ... }` to switch between the cases.

File: r/R/install-arrow.R [hunk 5]
```diff
@@ -56,7 +56,11 @@ install_arrow <- function(nightly = FALSE,
 
   if (sysname %in% c("windows", "darwin", "linux")) {
     if (conda && !nightly) {
```
Comment (nealrichardson) [inline r/R/install-arrow.R:58]:
```suggestion
    if (conda) {
```

**Review by nealrichardson — APPROVED** [review]
One suggestion in `install_arrow()` that I'll apply; otherwise LGTM!!

**Comment by bkietz** [comment]
JNI failure is unrelated, merging

**Pull request closed without merging**